README.md
# Yatzy Refactoring Kata

This Refactoring Kata was designed by Jon Jagger and is available in his Cyber-Dojo. See [his blog post](http://jonjagger.blogspot.co.uk/2012/05/yahtzee-cyber-dojo-refactoring-in-java.html)

I have changed it a little, so that the rules more closely match the original game.

The other language translations have been contributed by:

- Python: Emily Bache
- Ruby: Kim Persson and Lennart Fridén
- Javascript: Antti Tarvainen

## Kata: Yatzy rules

The game of Yatzy is a simple dice game. Each player
rolls five six-sided dice. They can re-roll some or all
of the dice up to three times (including the original roll).

For example, suppose a players rolls:

    3,4,5,5,2
    
They hold (-,-,5,5,-) and re-roll (3,4,-,-,2):

    5,1,5,5,3

They hold (5,-,5,5,-) and re-roll (-,1,-,-,3):

    5,6,5,5,2

The player then places the roll in a category, such as ones,
twos, fives, pair, two pairs etc (see below). If the roll is
compatible with the category, the player gets a score for the
roll according to the rules. If the roll is not compatible
with the category, the player scores zero for the roll.

For example, suppose a player scores 5,6,5,5,2 in the fives
category they would score 15 (three fives). The score for
that go is then added to their total and the category cannot
be used again in the remaining goes for that game. 
A full game consists of one go for each category. Thus, for
their last go in a game, a player must choose their only
remaining category.

Your task is to score a GIVEN roll in a GIVEN category.
You do NOT have to program the random dice rolling.
The game is NOT played by letting the computer choose the
highest scoring category for a given roll.
  

## Kata: Yatzy Categories and Scoring Rules

### Chance: 
The player scores the sum of all dice, no matter what they read.
For example:
  
-   1,1,3,3,6 placed on "chance" scores 14 (1+1+3+3+6)
-   4,5,5,6,1 placed on "chance" scores 21 (4+5+5+6+1)  

### Yatzy: 
If all dice have the same number,
the player scores 50 points. 
For example:
  
-   1,1,1,1,1 placed on "yatzy" scores 50
-   1,1,1,2,1 placed on "yatzy" scores 0

### Ones, Twos, Threes, Fours, Fives, Sixes: 
The player scores the sum of the dice that reads one, 
two, three, four, five or six, respectively. 
For example:

-   1,1,2,4,4 placed on "fours" scores 8 (4+4)
-   2,3,2,5,1 placed on "twos" scores 4  (2+2)
-   3,3,3,4,5 placed on "ones" scores 0

### Pair: 
The player scores the sum of the two highest matching dice.
For example, when placed on "pair":
  
-   3,3,3,4,4 scores 8 (4+4)
-   1,1,6,2,6 scores 12 (6+6)
-   3,3,3,4,1 scores 6 (3+3)
-   3,3,3,3,1 scores 6 (3+3)

### Two pairs: 
If there are two pairs of dice with the same number, the
player scores the sum of these dice. 
For example, when placed on "two pairs":
  
-   1,1,2,3,3 scores 8 (1+1+3+3)
-   1,1,2,3,4 scores 0
-   1,1,2,2,2 scores 6 (1+1+2+2)

### Three of a kind: 
If there are three dice with the same number, the player
scores the sum of these dice. 
For example, when placed on "three of a kind":
    
-    3,3,3,4,5 scores 9 (3+3+3)
-    3,3,4,5,6 scores 0
-    3,3,3,3,1 scores 9 (3+3+3)

### Four of a kind: 
If there are four dice with the same number, the player
scores the sum of these dice. 
For example, when placed on "four of a kind":
  
-    2,2,2,2,5 scores 8 (2+2+2+2)
-    2,2,2,5,5 scores 0
-    2,2,2,2,2 scores 8 (2+2+2+2)

### Small straight: 
When placed on "small straight", if the dice read

   1,2,3,4,5, 
   
the player scores 15 (the sum of all the dice).

### Large straight: 
When placed on "large straight", if the dice read

   2,3,4,5,6, 
   
the player scores 20 (the sum of all the dice).

### Full house: 
If the dice are two of a kind and three of a kind, the
player scores the sum of all the dice. 
For example, when placed on "full house":
   
-    1,1,2,2,2 scores 8 (1+1+2+2+2) 
-    2,2,3,3,4 scores 0
-    4,4,4,4,4 scores 0


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2] - 2019-10-18
### Added
- Improved tests coverage on `Yatzy` main class.

## [0.1] - 2019-10-17
### Added
- Initial pull from [Yatzy-Refactoring-Kata](https://github.com/emilybache/Yatzy-Refactoring-Kata).
- This CHANGELOG file to historize changes on source project.


src/test/java/bdd/YatzySteps.java
package bdd;

import cucumber.api.java.en.Given;
import cucumber.api.java.en.Then;
import cucumber.api.java.en.When;
import loicgeo.katas.yatzy.Yatzy;
import loicgeo.katas.yatzy.YatzyCategory;
import loicgeo.katas.yatzy.exception.FonctionalException;

import static java.lang.Character.getNumericValue;
import static org.assertj.core.api.Assertions.assertThat;

public class YatzySteps {
    private Yatzy yatzy;
    private YatzyCategory yatzyCategory;
    private int score;

    @Given("a dice serie {string}")
    public void aDiceSerie(String diceSerie) throws FonctionalException {
        String serie = diceSerie.replaceAll("[, ]", "");
        yatzy = new Yatzy(getValue(serie, 0), getValue(serie, 1), getValue(serie, 2), getValue(serie, 3), getValue(serie, 4));
    }

    private Integer getValue(String serie, int position) {
        return getNumericValue(serie.charAt(position));
    }

    @Given("a roll type {string}")
    public void aRollType(String rollType) {
        yatzyCategory = YatzyCategory.valueOf(rollType);
    }

    @When("Yatzy is scored")
    public void yatzyIsScored() {
        score = yatzy.scoreRoll(yatzyCategory);
    }

    @Then("the score is {int}")
    public void scoreShouldBe(int expectedScore) {
        assertThat(score).isEqualTo(expectedScore);
    }
}

src/test/java/bdd/RunCucumberTest.java
package bdd;

import cucumber.api.CucumberOptions;
import cucumber.api.junit.Cucumber;
import org.junit.runner.RunWith;

@RunWith(Cucumber.class)
@CucumberOptions(plugin = {"pretty", "html:target/bdd_results"})
public class RunCucumberTest {
}

src/test/java/loicgeo/katas/yatzy/YatzyTest.java
package loicgeo.katas.yatzy;

import junitparams.JUnitParamsRunner;
import junitparams.Parameters;
import loicgeo.katas.yatzy.exception.FonctionalException;
import org.assertj.core.api.ThrowableAssert;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.stream.Stream;

import static java.text.MessageFormat.format;
import static java.util.stream.Collectors.joining;
import static loicgeo.katas.yatzy.YatzyCategory.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@RunWith(JUnitParamsRunner.class)
public class YatzyTest {

    @Test
    public void should_throw_an_functional_exception_for_a_wrong_dice_serie_scoring() {
        // given
        ThrowableAssert.ThrowingCallable yatzyInstantiation = () -> new Yatzy(0, 3, 4, 5, 1);

        // when
        // then
        assertThatThrownBy(yatzyInstantiation)
                .isInstanceOf(FonctionalException.class)
                .hasMessageContaining("Wrong dice values");
    }

    @Test
    public void should_score_a_dice_serie_for_a_chance_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 4, 5, 1);
        // when
        int score = yatzy.scoreRoll(CHANCE);
        // then
        assertThat(score).isEqualTo(16);
    }

    @Test
    public void should_score_50_a_dice_serie_for_a_yatzy_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(4, 4, 4, 4, 4);
        // when
        int score = yatzy.scoreRoll(YATZY);
        // then
        assertThat(score).isEqualTo(50);
    }

    @Test
    public void should_score_0_a_dice_serie_for_a_non_yatzy_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(6, 6, 6, 6, 3);
        // when
        int score = yatzy.scoreRoll(YATZY);
        // then
        assertThat(score).isEqualTo(0);
    }

    public static Object[] givenSeriesForEachKindOfValues() {
        return new Object[]{
                new Object[]{1, 2, 3, 4, 5, ONES, 1},
                new Object[]{1, 2, 1, 4, 5, ONES, 2},
                new Object[]{6, 2, 2, 4, 5, ONES, 0},
                new Object[]{1, 1, 1, 1, 1, ONES, 5},

                new Object[]{1, 2, 3, 2, 6, TWOS, 4},
                new Object[]{2, 2, 2, 2, 2, TWOS, 10},

                new Object[]{1, 2, 3, 2, 3, THREES, 6},
                new Object[]{3, 3, 3, 3, 3, THREES, 15},

                new Object[]{4, 4, 4, 5, 5, FOURS, 12},
                new Object[]{4, 4, 5, 5, 5, FOURS, 8},
                new Object[]{4, 5, 5, 5, 5, FOURS, 4},

                new Object[]{4, 4, 4, 5, 5, FIVES, 10},
                new Object[]{4, 4, 5, 5, 5, FIVES, 15},
                new Object[]{4, 5, 5, 5, 5, FIVES, 20},

                new Object[]{4, 4, 4, 5, 5, SIXES, 0},
                new Object[]{4, 4, 6, 4, 5, SIXES, 6},
                new Object[]{6, 5, 6, 6, 5, SIXES, 18}
        };
    }

    @Test
    @Parameters(method = "givenSeriesForEachKindOfValues")
    public void should_score_several_dice_series_for_each_kind_of_value(int d1, int d2, int d3, int d4, int d5, YatzyCategory yatzyCategory, int expectedScore) throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(d1, d2, d3, d4, d5);
        // when
        int score = yatzy.scoreRoll(yatzyCategory);
        // then
        assertThat(score)
                .overridingErrorMessage(format("A {0} roll should score {1} for values ''{2}''",
                        yatzyCategory,
                        expectedScore,
                        Stream.of(d1, d2, d3, d4, d5).map(Object::toString).collect(joining(", "))))
                .isEqualTo(expectedScore)
        ;
    }

    @Test
    public void should_score_dice_serie_for_pair_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 4, 3, 5, 6);
        // when
        int score = yatzy.scoreRoll(PAIR);
        // then
        assertThat(score).isEqualTo(6);
    }

    @Test
    public void should_score_0_for_pair_roll_having_none_pair() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(1, 3, 4, 6, 5);
        // when
        int score = yatzy.scoreRoll(PAIR);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_score_the_best_pair_for_pair_roll_having_2_pairs() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(5, 3, 6, 6, 5);
        // when
        int score = yatzy.scoreRoll(PAIR);
        // then
        assertThat(score).isEqualTo(12);
    }

    @Test
    public void should_score_only_2_values_for_a_pair_roll_composed_of_3_same_values() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 3, 4, 1);
        // when
        int score = yatzy.scoreRoll(PAIR);
        // then
        assertThat(score).isEqualTo(6);
    }

    @Test
    public void should_score_a_2_pairs_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 5, 4, 5);
        // when
        int score = yatzy.scoreRoll(TWO_PAIRS);
        // then
        assertThat(score).isEqualTo(16);
    }

    @Test
    public void should_score_a_2_pairs_roll_composed_of_3_same_values() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 5, 5, 5);
        // when
        int score = yatzy.scoreRoll(TWO_PAIRS);
        // then
        assertThat(score).isEqualTo(16);
    }

    @Test
    public void should_score_0_a_2_pairs_roll_without_2_pairs() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(1, 1, 2, 3, 4);
        // when
        int score = yatzy.scoreRoll(TWO_PAIRS);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_score_three_values_of_a_kind() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 3, 4, 5);
        // when
        int score = yatzy.scoreRoll(THREE_Of_A_KIND);
        // then
        assertThat(score).isEqualTo(9);
    }

    @Test
    public void should_score_0_for_none_match_three_values_of_a_kind() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(1, 3, 5, 4, 5);
        // when
        int score = yatzy.scoreRoll(THREE_Of_A_KIND);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_only_3_values_of_a_4_occurrenced_values_for_a_three_values_of_a_kind() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 3, 3, 5);
        // when
        int score = yatzy.scoreRoll(THREE_Of_A_KIND);
        // then
        assertThat(score).isEqualTo(9);
    }

    @Test
    public void should_score_four_values_of_a_kind() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 3, 3, 3, 5);
        // when
        int score = yatzy.scoreRoll(FOUR_Of_A_KIND);
        // then
        assertThat(score).isEqualTo(12);
    }

    @Test
    public void should_score_0_for_none_match_four_values_of_a_kind() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(3, 1, 2, 3, 3);
        // when
        int score = yatzy.scoreRoll(FOUR_Of_A_KIND);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_only_4_values_of_a_5_occurrenced_values_for_a_three_values_of_a_kind() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(2, 2, 2, 2, 2);
        // when
        int score = yatzy.scoreRoll(FOUR_Of_A_KIND);
        // then
        assertThat(score).isEqualTo(8);
    }

    @Test
    public void should_sum_values_of_a_small_straight() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(2, 3, 4, 5, 1);
        // when
        int score = yatzy.scoreRoll(SMALL_STRAIGHT);
        // then
        assertThat(score).isEqualTo(15);
    }

    @Test
    public void should_sum_0_for_non_existing_small_straight() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(1, 2, 2, 4, 5);
        // when
        int score = yatzy.scoreRoll(SMALL_STRAIGHT);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_0_on_large_straight_for_a_small_straight_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(2, 3, 4, 5, 6);
        // when
        int score = yatzy.scoreRoll(SMALL_STRAIGHT);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_values_of_a_large_straight() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(6, 2, 3, 4, 5);
        // when
        int score = yatzy.scoreRoll(LARGE_STRAIGHT);
        // then
        assertThat(score).isEqualTo(20);
    }

    @Test
    public void should_sum_0_for_non_existing_large_straight() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(1, 2, 2, 4, 5);
        // when
        int score = yatzy.scoreRoll(LARGE_STRAIGHT);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_0_on_small_straight_for_a_large_straight_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(1, 2, 3, 4, 5);
        // when
        int score = yatzy.scoreRoll(LARGE_STRAIGHT);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_a_full_house_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(6, 2, 2, 2, 6);
        // when
        int score = yatzy.scoreRoll(FULL_HOUSE);
        // then
        assertThat(score).isEqualTo(18);
    }

    @Test
    public void should_sum_0_a_non_full_house_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(2, 3, 4, 5, 6);
        // when
        int score = yatzy.scoreRoll(FULL_HOUSE);
        // then
        assertThat(score).isEqualTo(0);
    }

    @Test
    public void should_sum_0_a_yatzy_for_a_full_house_roll() throws FonctionalException {
        // given
        Yatzy yatzy = new Yatzy(2, 2, 2, 2, 2);
        // when
        int score = yatzy.scoreRoll(FULL_HOUSE);
        // then
        assertThat(score).isEqualTo(0);
    }
    
}


src/main/java/loicgeo/katas/yatzy/Yatzy.java
package loicgeo.katas.yatzy;

import loicgeo.katas.yatzy.exception.FonctionalException;

import java.util.logging.Level;
import java.util.logging.Logger;

public class Yatzy {

    private static final Logger LOGGER = Logger.getLogger(Yatzy.class.getName());

    private DiceSerie diceSerie;

    /**
     * Instantiation a Yatzy dice serie.
     *
     * @param d1 number value for dice #1
     * @param d2 number value for dice #2
     * @param d3 number value for dice #3
     * @param d4 number value for dice #4
     * @param d5 number value for dice #5
     * @throws FonctionalException in case of invalid dice serie. See details on {@link DiceSerie}.
     * @see DiceSerie
     * @since 0.2
     */
    public Yatzy(int d1, int d2, int d3, int d4, int d5) throws FonctionalException {
        try {
            diceSerie = new DiceSerie(d1, d2, d3, d4, d5);
        } catch (FonctionalException exception) {
            LOGGER.log(Level.SEVERE, "Unable to instantiate Yatzy", exception);
            throw exception;
        }
    }

    public int scoreRoll(YatzyCategory category) {
        return category.score(diceSerie);
    }

}





src/main/java/loicgeo/katas/yatzy/DiceSerie.java
package loicgeo.katas.yatzy;

import loicgeo.katas.yatzy.exception.FonctionalException;

import java.util.Collection;
import java.util.Map;
import java.util.function.Predicate;

import static java.text.MessageFormat.format;
import static java.util.Arrays.asList;
import static java.util.function.Function.identity;
import static java.util.stream.Collectors.*;

public class DiceSerie {

    private static final int MIN_DICE_VALUE = 1;
    private static final int MAX_DICE_VALUE = 6;

    private static final Predicate<? super Integer> NOT_ALLOWED_VALUE = value -> value < MIN_DICE_VALUE || value > MAX_DICE_VALUE;

    private Collection<Integer> values;

    /**
     * Initialize a dice serie of 5 int values.
     *
     * @param value1 value dice 1
     * @param value2 value dice 2
     * @param value3 value dice 3
     * @param value4 value dice 4
     * @param value5 value dice 5
     * @throws FonctionalException in case of not allowed values.
     */
    public DiceSerie(int value1, int value2, int value3, int value4, int value5) throws FonctionalException {
        values = asList(value1, value2, value3, value4, value5);
        validateValues();
    }

    private void validateValues() throws FonctionalException {
        if (values.stream().anyMatch(NOT_ALLOWED_VALUE)) {
            throw new FonctionalException(
                    format("Wrong dice values: a dice serie must be composed of values between {0} and {1} included.",
                            MIN_DICE_VALUE,
                            MAX_DICE_VALUE
                    ));
        }
    }

    /**
     * Get number of dices of a value, for each possible value.
     * For example:
     * <li>for dices serie `1, 2, 3, 4, 5`, returns array int[]{0, 1, 1, 1, 1, 1, 0}</li>
     * <li>for dices serie `2, 2, 2, 5, 5`, returns array int[]{0, 0, 3, 0, 0, 2, 0}</li>
     *
     * @return an array of length MAX_DICE_VALUE + 1 (0 is ignored), composed of the number of dices having a value, for each value (i.e. the array indexes).
     */
    public int[] getNbOccurrenciesByValue() {
        int[] countsByValue = new int[MAX_DICE_VALUE + 1];

        Map<Integer, Long> countsByValueMap = values.stream().collect(groupingBy(identity(), counting()));
        countsByValueMap.forEach((value, count) -> countsByValue[value] = count.intValue());

        return countsByValue;
    }

    public boolean isComposedOfOnlyOneKindOfValue() {
        return values.stream().distinct().count() == 1;
    }

    public String getValuesAsString(String delimiter) {
        return values.stream()
                .sorted()
                .map(Object::toString)
                .collect(joining(delimiter));
    }

    public Collection<Integer> getValues() {
        return values;
    }
}


src/main/java/loicgeo/katas/yatzy/YatzyCategory.java
package loicgeo.katas.yatzy;

import java.util.function.ToIntFunction;

import static loicgeo.katas.yatzy.YatzyCategory.DefaultScores.*;

public enum YatzyCategory {

    // sum all values
    CHANCE(dices -> dices.getValues().stream().reduce(0, Integer::sum)),

    // only one distinct value
    YATZY(dices -> dices.isComposedOfOnlyOneKindOfValue() ? SCORE_YATZY.getScore() : 0),

    // sum all dice by value
    ONES(dices -> sumByDiceValue(dices, 1)),
    TWOS(dices -> sumByDiceValue(dices, 2)),
    THREES(dices -> sumByDiceValue(dices, 3)),
    FOURS(dices -> sumByDiceValue(dices, 4)),
    FIVES(dices -> sumByDiceValue(dices, 5)),
    SIXES(dices -> sumByDiceValue(dices, 6)),

    // sum values for the best pair
    PAIR(dices -> scoreValues(dices, 2, 1)),
    // sum values for the best 2 pairs
    TWO_PAIRS(dices -> scoreValues(dices, 2, 2)),

    // sum values for value having 3 occurrences
    THREE_Of_A_KIND(dices -> scoreValues(dices, 3, 1)),
    // sum values for value having 4 occurrences
    FOUR_Of_A_KIND(dices -> scoreValues(dices, 4, 1)),

    // sum values for a small straight, meaning composed of a continuous sequence of at least 4 values, starting by 1
    SMALL_STRAIGHT(dices -> dices.getValuesAsString("").contains("12345") ? SCORE_SMALL_STRAIGHT.getScore() : 0),
    // sum values for a large straight, meaning composed of a continuous sequence of at least 4 values, finishing by 6
    LARGE_STRAIGHT(dices -> dices.getValuesAsString("").contains("23456") ? SCORE_LARGE_STRAIGHT.getScore() : 0),

    // sum values for a set of 3 values X and a couple of values Y, where X is different of Y
    FULL_HOUSE(dices -> {
        if (dices.isComposedOfOnlyOneKindOfValue()) {
            return 0;
        }
        return scoreValues(dices, 3, 1)
                + scoreValues(dices, 2, 1);
    });

    private final ToIntFunction<DiceSerie> scoringFunction;

    YatzyCategory(ToIntFunction<DiceSerie> scoringFunction) {
        this.scoringFunction = scoringFunction;
    }

    public int score(DiceSerie diceSerie) {
        return scoringFunction.applyAsInt(diceSerie);
    }

    /**
     * Score a dice serie according:
     * <li>the number of occurrences for a value</li>
     * <li>the number of times this match is repeated</li>
     *
     * @param diceSerie                    a set of dice values
     * @param havingNbOccurrencesForAValue number of occurrences to find any value
     * @param nbTimesIntheSerie            number to match this motif (ex: for a pair, 2 pairs, etc)
     * @return
     */
    private static int scoreValues(DiceSerie diceSerie, final int havingNbOccurrencesForAValue, final int nbTimesIntheSerie) {
        int[] countsByValue = diceSerie.getNbOccurrenciesByValue();

        int score = 0;
        int nbMatchs = 0;
        for (int currentValue = countsByValue.length - 1; currentValue > 0; currentValue--) {
            int nbOccurrencesOfCurrentValue = countsByValue[currentValue];
            if (nbOccurrencesOfCurrentValue >= havingNbOccurrencesForAValue) {
                nbMatchs++;
                score += currentValue * Math.min(nbOccurrencesOfCurrentValue, havingNbOccurrencesForAValue);
                if (nbMatchs >= nbTimesIntheSerie) {
                    break;
                }
            }
        }
        return nbTimesIntheSerie == nbMatchs ? score : 0;
    }

    private static Integer sumByDiceValue(DiceSerie diceSerie, int value) {
        return diceSerie.getValues().stream()
                .filter(v -> v == value)
                .reduce(0, Integer::sum);
    }

    protected enum DefaultScores {
        SCORE_YATZY(50),
        SCORE_SMALL_STRAIGHT(15),
        SCORE_LARGE_STRAIGHT(20);

        private int score;

        DefaultScores(int defaultScore) {
            this.score = defaultScore;
        }

        public int getScore() {
            return score;
        }
    }

}
